File: gnsals0904/level1/옹알이2.java
class Solution {
    public int solution(String[] babbling) {
        int answer = 0;

        String[] words = {"aya", "ye", "woo", "ma"};

        for (String str : babbling) {
            int index = 0;
            String prev = "";
            boolean possible = true;

            while (index < str.length()) {
                boolean found = false;

                for (String word : words) {
                    if (str.startsWith(word, index) && !word.equals(prev)) {
                        index += word.length();
                        prev = word;
                        found = true;
                        break;
                    }
                }

                if (!found) {
                    possible = false;
                    break;
                }
            }

            if (possible) {
                answer++;
            }
        }

        return answer;
    }
}

File: gpfksdlrn/level1/옹알이_2.js
// https://school.programmers.co.kr/learn/courses/30/lessons/133499

function solution(babbling) {
  const sounds = ['aya', 'ye', 'woo', 'ma'];

  const canSay = (word) => {
    let i = 0;
    let prev = '';

    while (i < word.length) {
      const sound = sounds.find((s) => s !== prev && word.startsWith(s, i));
      if (!sound) return false;

      i += sound.length;
      prev = sound;
    }

    return true;
  };

  return babbling.filter(canSay).length;
}
